ssl/ech/ech_internal.c

```diff
@@ -160,4 +226,1251 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+        return 0;
+    }
+    if (!WPACKET_put_bytes_u16(pkt, s->ext.ech.attempted_type)
+        || !WPACKET_start_sub_packet_u16(pkt)
+        || !WPACKET_put_bytes_u8(pkt, OSSL_ECH_OUTER_CH_TYPE)
+        || !WPACKET_put_bytes_u16(pkt, hpke_suite.kdf_id)
+        || !WPACKET_put_bytes_u16(pkt, hpke_suite.aead_id)
+        || !WPACKET_memcpy(pkt, &cid, 1)
+        || !WPACKET_sub_memcpy_u16(pkt, senderpub, senderpub_len)
+        || !WPACKET_sub_memcpy_u16(pkt, cipher, cipher_len)
+        || !WPACKET_close(pkt)
+        ) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /* record the ECH sent so we can re-tx same if we hit an HRR */
+    OPENSSL_free(s->ext.ech.sent);
+    WPACKET_get_total_written(pkt, &pp_at_end);
+    s->ext.ech.sent_len = pp_at_end - pp_at_start;
+    s->ext.ech.sent = OPENSSL_malloc(s->ext.ech.sent_len);
+    if (s->ext.ech.sent == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    memcpy(s->ext.ech.sent, pp, s->ext.ech.sent_len);
+    s->ext.ech.grease = OSSL_ECH_IS_GREASE;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "ECH - sending GREASE\n");
+    } OSSL_TRACE_END(TLS);
+    return 1;
+}
+
+/*
+ * Search the ECH store for one that's a match. If no outer_name was set via
+ * API then we just take the 1st match where we locally support the HPKE suite.
+ * If OTOH, an outer_name was provided via API then we prefer the first that
+ * matches that. Name comparison is via case-insensitive exact matches.
+ */
+int ossl_ech_pick_matching_cfg(SSL_CONNECTION *s, OSSL_ECHSTORE_ENTRY **ee,
+                               OSSL_HPKE_SUITE *suite)
+{
+    int namematch = 0, nameoverride = 0, suitematch = 0, num, cind = 0;
+    unsigned int csuite = 0, hnlen = 0;
+    OSSL_ECHSTORE_ENTRY *lee = NULL;
+    OSSL_ECHSTORE *es = NULL;
+    char *hn = NULL;
+
+    if (s == NULL || s->ext.ech.es == NULL || ee == NULL || suite == NULL)
+        return 0;
+    es = s->ext.ech.es;
+    if (es->entries == NULL)
+        return 0;
+    num = sk_OSSL_ECHSTORE_ENTRY_num(es->entries);
+    /* allow API-set pref to override */
+    hn = s->ext.ech.outer_hostname;
+    hnlen = (hn == NULL ? 0 : strlen(hn));
+    if (hnlen != 0)
+        nameoverride = 1;
+    if (s->ext.ech.no_outer == 1) {
+        hn = NULL;
+        hnlen = 0;
+        nameoverride = 1;
+    }
+    for (cind = 0; cind != num && (suitematch == 0 || namematch == 0); cind++) {
+        lee = sk_OSSL_ECHSTORE_ENTRY_value(es->entries, cind);
+        if (lee == NULL || lee->version != OSSL_ECH_RFCXXXX_VERSION)
+            continue;
+        if (nameoverride == 1) {
+            namematch = 1;
+        } else {
+            namematch = 0;
+            if (hnlen == 0
+                || (lee->public_name != NULL
+                    && strlen(lee->public_name) == hnlen
+                    && !OPENSSL_strncasecmp(hn, (char *)lee->public_name,
```

Comment:
We'll hit that case if `outer_hostname` was not set and `no_outer` was not set, which is the typical case. That means `hnlen` will be zero and we shouldn't compare to the `public_name` from the ECHConfig but just accept whatever's in the ECHconfig so long as that has a `public_name`

OTOH, if `outer_hostname` was set (but `no_outer` was not), then we do want to do the string match.

Reply:
> OTOH, if outer_hostname was set (but no_outer was not), then we do want to do the string match.

But the code as written doesn't seem to do the string match in that case? If `outer_hostname` was set then `hnlen > 0`. In that case line 384 sets `nameoverride` to 1. That means we never get to this string match because the test on line 394 evaluates to true, and we never enter this else branch.

Reply:
sorry, yes, I misspoke in that last, as you say, the code is currently skipping the name comparison if we got an override, looking at it now...

Reply:
OK, I've tweaked to fix - looks like something did get lost in translation there somewhere along the line, but there are test cases to add for this that I'll include in the  next PR - testing this stuff needs both client and server sides, so those tests aren't included in this PR.